insteon_mqtt/device/functions/Set.py

```diff
@@ -0,0 +1,81 @@
+#===========================================================================
+#
+# Set Functions.
+#
+#===========================================================================
+from ..Base import Base
+from ... import log
+from ... import on_off
+
+LOG = log.get_logger()
+
+
+class Set(Base):
+    """Scene Trait Abstract Class
+
+    This is an abstract class that provides support for the Scene topic.
+    """
+    def __init__(self, protocol, modem, address, name=None):
+        """Constructor
+
+        Args:
+          protocol (Protocol): The Protocol object used to communicate
+                   with the Insteon network.  This is needed to allow the
+                   device to send messages to the PLM modem.
+          modem (Modem): The Insteon modem used to find other devices.
+          address (Address): The address of the device.
+          name (str): Nice alias name to use for the device.
+        """
+        super().__init__(protocol, modem, address, name)
+
+        self.cmd_map.update({
+            'set' : self.set,
+            })
+
+    #-----------------------------------------------------------------------
+    def set(self, is_on=None, level=None, group=0x01, mode=on_off.Mode.NORMAL,
+            reason="", transition=None, on_done=None):
+        """Turn the device on or off.  Level zero will be off.
+
+        NOTE: This does NOT simulate a button press on the device - it just
+        changes the state of the device.  It will not trigger any responders
+        that are linked to this device.  To simulate a button press, call the
+        scene() method.
+
+        This will send the command to the device to update it's state.  When
+        we get an ACK of the result, we'll change our internal state and emit
+        the state changed signals.
+
+        Args:
+          is_on (bool): True to turn on, False for off
+          level (int): If non zero, turn the device on.  Should be in the
+                range 0 to 255.  If None, use default on-level.
+          group (int): The group to send the command to.  For this device,
+                this must be 1.  Allowing a group here gives us a consistent
+                API to the on command across devices.
+          mode (on_off.Mode): The type of command to send (normal, fast, etc).
+          reason (str):  This is optional and is used to identify why the
+                 command was sent. It is passed through to the output signal
+                 when the state changes - nothing else is done with it.
+          transition (int): The transition ramp_rate if supported.
+          on_done: Finished callback.  This is called when the command has
+                   completed.  Signature is: on_done(success, msg, data)
+        """
+        if is_on or level:
```

Comment:
Well, in #323, the template did end up computing {'cmd': 'off', 'level': 255} - basically because the old default template from versions prior to 0.7.5 adds `"level" : 255` to all commands if brightness isn't present.

From #323:
> I found it interesting that the command was calculated as {'cmd': 'off', 'level': 255}. Off should be associated with level 0, correct? Anyway, everything works and the published command was { "state" : "OFF", "brightness" : 0 }, so it all worked out. I still found it interesting. I guess it's not a bug if it works.

So, if level takes priority with the updated code, then that would represent a change in behavior - at least for installations that have templates that add `"level" : 255` to commands where brightness isn't present

Reply:
Dimmer Level Template
```yaml
    level_payload: >
       { "cmd" : "{{json.state.lower()}}",
         "level" : {% if json.brightness is defined %}
                      {{json.brightness}}
                   {% else %}
                      255
                   {% endif %} }
```

```
(venv) krkeegan@batman:~/projects/insteon-mqtt$ git checkout 0.7.5
HEAD is now at 9630b9f Merge pull request #298 from TD22057/dev
(venv) krkeegan@batman:~/projects/insteon-mqtt$ mosquitto_pub -t insteon/4f.23.38/level -m '{"state":"OFF"}'
```
Log
```
2021-01-15 13:35:59.620 INFO Dimmer: Dimmer message insteon/4f.23.38/level b'{"state":"OFF"}'
2021-01-15 13:35:59.620 DEBUG MsgTemplate: Input template render: '{ "cmd" : "off",
  "level" : 
               255
             }'
2021-01-15 13:35:59.620 INFO Dimmer: Dimmer input command: {'cmd': 'off', 'level': 255}
2021-01-15 13:35:59.620 INFO Dimmer: Dimmer 4f.23.38 cmd: off
2021-01-15 13:35:59.620 DEBUG MsgHistory: Average hops 0.0, using 0
2021-01-15 13:35:59.620 INFO Protocol: Write message to modem: Std: 4f.23.38, Type.DIRECT mh:0 hl:0, 13 00 
```

```
(venv) krkeegan@batman:~/projects/insteon-mqtt$ git checkout Set_Topic 
Switched to branch 'Set_Topic'
Your branch is up to date with 'origin/Set_Topic'.
(venv) krkeegan@batman:~/projects/insteon-mqtt$ mosquitto_pub -t insteon/4f.23.38/level -m '{"state":"OFF"}'
```

LOG

```
2021-01-15 13:39:24.026 INFO Dimmer: Dimmer message insteon/4f.23.38/level b'{"state":"OFF"}'
2021-01-15 13:39:24.026 DEBUG MsgTemplate: Input template render: '{ "cmd" : "off",
  "level" : 
               255
             }'
2021-01-15 13:39:24.027 INFO Dimmer: Dimmer input command: {'cmd': 'off', 'level': 255}
2021-01-15 13:39:24.027 INFO Dimmer: Dimmer 4f.23.38 cmd: off
2021-01-15 13:39:24.027 INFO Protocol: Write message to modem: Std: 4f.23.38, Type.DIRECT mh:3 hl:3, 13 00 
```

The level topic still cleans up for this mistake in the config payload topic.   What am I missing?

Reply:
Maybe nothing.  Upon a closer look, my initial "maybe this isn't a concern" may be the case.

> Maybe this isn't a concern, because level is never provided to the on/off topics?

In particular, the level topic has `level = '0' if not is_on else data.get('level', None)` and the on-off topic will never pass a level unless someone changes the on_off_payload to pass a level like the old level topic used to do.  (Currently, the default on-off template (`{ "cmd" : "{{value.lower()}}" }`) will only pass a "cmd" value, though.

If you want to put a belt and suspenders on it, you could change _input_set() in mqtt/topic/SetTopic.py to do  `level = '0' if not is_on else data.get('level', None)` instead of `level = data.get("level", None)`.  That would ensure that, even if someone changed the on_off_payload to be capable of passing `{'cmd': 'off', 'level': 255}`, the `'cmd': 'off'` portion would override.

I'd also be fine with leaving it alone, since this can only cause confusion if someone messes with the on_off_payload.

Reply:
Yeah, it is all a bit of a mess, see for example
```
(venv) krkeegan@batman:~/projects/insteon-mqtt$ git checkout 0.7.5
HEAD is now at 9630b9f Merge pull request #298 from TD22057/dev
(venv) krkeegan@batman:~/projects/insteon-mqtt$ ./insteon-mqtt config/config.yaml on dimmer --level=0
Commanding dimmer device 4f.23.38 (dimmer) cmd=on
Dimmer state updated to 0
```
This still works in this branch as well.  The issue really is issuing two conflicting commands is always ambiguous, we have to chose one.  I agree the level topic needs to catch `level=255 and is_on=False` scenario since we apparently created that confusion ourselves.  I am less convinced that we need to go and edit the `Set_Topic` to catch the same error, since as you note a user would have to proactively do that to themselves, and even then we would be resolving their ambiguity for them.

I say leave it, unless it bothers you.

Reply:
Sounds good.  Let's leave it, then.